### PR TITLE
Random alternate sprite graphics [WIP]

### LINKED ISSUE #11867: Allow multiple versions of the same tile? ($30 on Bountysource)
While implementing #11823 it occurred to me that it would be nice to have multiple versions of the same tile. Walls with graffiti, lopsided trees, etc.

Simply replacing each sprite number with an array of sprite numbers to be chosen between is a simple idea, but perhaps too simple. Some alternatives should be more common than others. Maybe we could just hard code a probability curve? first sprite 80%, second sprite 16%, etc.

Also, keeping the view consistent would be important. I'm thinking the random selection could be based on the global X/Y/Z coordinates of the tile. Are there any non-cryptographic hashes that are fast enough to put inside the tile drawing loop that would produce results that don't appear patterned to a human if they just lay down a solid block of those tiles?

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/10279028-allow-multiple-versions-of-the-same-tile?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by sparr**
This patch allows the sprite id or list in a tileset to be replaced by an object describing multiple weighted variations. The proof-of-concept tileset is Retrodays10px_test which uses the existing Retrodays10px graphics and simply specifies an existing sprite (a grey "ladder" graphic, sprite number 36) as a 10% alternate for the grass tile as follows:

```
    {
        "id":"t_grass",
        "fg": [
            { "weight":9, "sprite":5 },
            { "weight":1, "sprite":36 }
        ],
        "rotates":false
    },
```

This patch is not complete yet. The following concerns remain:

The use of simplexnoise.h:raw_noise_2d is both slower than I would like and not uniformly distributed (notice that the 10% ladder graphic only appears approx 4% of the time). An alternate method for choosing a variant consistently for a given tile would resolve both of these problems.

I have not been able to test every edge case of variations of tile configuration yet.

Also I used a slightly inefficient solution for the variant selection. The bookkeeping vector allows O(1) lookups but uses O(sum_of_weights) memory. An implementation of the alias method would be a slightly slower O(1) for lookups, and take only O(number_of_variants) memory. This will only be important if tileset authors define very large weights (weights of 999 and 1 are required for a variant to appear only 0.1% of the time).

A tileset with real variant sprites would be a good thing to include with deployment of this patch.

It may be possible or desirable to replace the existing id_or_id functionality used to randomly choose between dirt and grass during map generation with a use of this functionality instead.

Feedback on the functionality in this patch so far would be appreciated, as would ideas on solving or ignoring the above concerns.

Closes #11867

### CODE REVIEW

**Comment by Zireael07** [comment]
Random dirt or grass sounds great!

File: src/cata_tiles.h [hunk 1]
```diff
@@ -25,9 +25,37 @@ class JsonObject;
...
+    {
+        weight = 1;
+        sprite_ids.clear();
+        sprite_ids.push_back(id);
+    }
+
+    sprite_variation(int w, int id)
+    {
+        weight = w;
+        sprite_ids.clear();
```
Comment (BevapDin) [inline src/cata_tiles.h]: Why all those `clear` calls in the constructors? Do you think the vector is initialized to some content?

To make this clear: a vector is initialized by its default constructor to be empty. The `clear` calls here a re completely pointless.

And why not use the `weighted_list` class from src/weighted_list.h?

File: src/cata_tiles.h [hunk 2]
```diff
@@ -25,9 +25,37 @@ class JsonObject;
 struct visibility_variables;
 
 /** Structures */
+struct sprite_variation {
+    int weight;
+    std::vector<int> sprite_ids;
+
+    sprite_variation()
+    {
+        weight = 1;
```
Comment (BevapDin) [inline src/cata_tiles.h]: Why does the empty list start with a weight of 1? In line 723, you explicitly set the weight to -1, directly after construction a `sprite_variation`. This is very confusing. Apparently default  constructed objects of this type are not functional and need their `weight` member set to -1 manually, or not? If so, it should be set in the constructor (here).

File: src/cata_tiles.h (hunk 1 above)
Comment (sparr) [inline src/cata_tiles.h]: clear calls copied from other nearby structs with vector initializers. so, ignorance.

I'll definitely use weighted_list now that I know about it :)

File: src/cata_tiles.cpp [hunk 3]
```diff
@@ -680,43 +716,72 @@ tile_type &cata_tiles::load_tile(JsonObject &entry, const std::string &id, int o
...
+            // create one variation per object
+            else if ( g_array.test_object() ) {
+                while (g_array.has_more()) {
+                    sprite_variation v = sprite_variation();
+                    JsonObject vo = g_array.next_object();
+                    v.weight = vo.get_int("weight");
+                    // negative weight is invalid
+                    if ( v.weight < 0 ) {
+                        vo.throw_error("Invalid weight for sprite variation (<0)", g.first);
+                        continue;
```
Comment (BevapDin) [inline src/cata_tiles.cpp]: The continue here is never reached. `throw_error` throws an exception and never returns normally. This is like:
```C++
if( ... ) {
    throw JsonError( "error" );
    continue; // neve rreached.
}
```

File: src/cata_tiles.h (hunk 1 above)
Comment (sparr) [inline src/cata_tiles.h]: scratch that. weighted_list is good to know about, and I might use it, however I would have to improve it. weighted_list.pick_ent() is O(N). that's not acceptable for something potentially called for every single tile being drawn. I might add optional precalculation to it.

File: src/cata_tiles.cpp [hunk 4]
```diff
@@ -680,43 +716,72 @@ tile_type &cata_tiles::load_tile(JsonObject &entry, const std::string &id, int o
...
+                        while (sprites.has_more()) {
+                            const int s = sprites.next_int();
+                            if ( s >= 0 && s < size )
+                            {
+                                v.sprite_ids.push_back(s);
+                            }
+                            else
+                            {
+                                vo.throw_error("Invalid value for sprite id (out of range)", g.first);
+                                v.sprite_ids.push_back(s+offset);
```
Comment (BevapDin) [inline src/cata_tiles.cpp]: Same as  above, this line is never reached.

File: src/cata_tiles.h (hunk 2 above)
Comment (sparr) [inline src/cata_tiles.h]: the default weight of 1 is functional for every case except the first one. on line 723 I'm creating an "invalid" entry for storing the O(1) lookup list of other variations. every other instance of a sprite_variation uses a positive weight, and most of them use the default weight of 1.

File: src/cata_tiles.cpp (hunk 3 above)
Comment (sparr) [inline src/cata_tiles.cpp]: thanks. I'll need to use something other than throw_error for cases where I want to display a warning in context but still continue execution.

File: src/cata_tiles.cpp [hunk 5]
```diff
@@ -680,43 +716,72 @@ tile_type &cata_tiles::load_tile(JsonObject &entry, const std::string &id, int o
...
     tile_type &curr_subtile = tile_ids[id];
 
-    if ( entry.has_array("fg") ) {
-        JsonArray fg_array = entry.get_array("fg");
-        while (fg_array.has_more()) {
-            const int fg = fg_array.next_int();
-            if ( fg >= 0 ) {
-                curr_subtile.fg.push_back(fg);
+    // loop through fg and bg
+    for (auto g : {std::make_pair("fg",&curr_subtile.fg), std::make_pair("bg",&curr_subtile.bg)}) {
```
Comment (BevapDin) [inline src/cata_tiles.cpp]: Instead of a loop (which is clever, but confusing), put the body of the loop into a separate function and call it once for `fg` and once for `bg`.

**Comment by chaosvolt** [comment]
Hmm. Will be interesting, though pondering how best to make use of it. X3

**Comment by sparr** [comment]
@chaosvolt some of the ideas from when this was proposed: graffiti on walls, cracks in pavement, multiple types of trees, blood on walls.

File: src/cata_tiles.cpp [hunk 6]
```diff
@@ -502,14 +503,49 @@ void cata_tiles::load_tilejson_from_file(const std::string &tileset_dir, std::if
...
+    // loop through all tile ids
     for( auto it = tile_ids.begin(); it != tile_ids.end(); ) {
+        // second is the tile_type describing that id
         auto &td = it->second;
-        td.fg.erase(std::remove_if(td.fg.begin(), td.fg.end(),
-                               [&](int i) { return i >= offset || i < 0; }), td.fg.end());
-        td.bg.erase(std::remove_if(td.bg.begin(), td.bg.end(),
-                               [&](int i) { return i >= offset || i < 0; }), td.bg.end());
+        // loop through fg and bg
+        for (auto *g : {&td.fg, &td.bg}) {
```
Comment (BevapDin) [inline src/cata_tiles.cpp]: Same here. It probably works fine as it is, but it is much more confusing than a simple function call.

**Comment by sparr** [comment]
just-thought-of bug: if this feature is applied to vehicle part sprites, they will change as the vehicle moves. need to either exclude movable things, or use different random seeds for vehicles, furniture, etc.

File: src/cata_tiles.cpp [hunk 7]
```diff
@@ -502,14 +503,49 @@ void cata_tiles::load_tilejson_from_file(const std::string &tileset_dir, std::if
...
+            // loop through all of the variations
+            for( auto vit = g->begin(); vit != g->end(); vit++) {
```
Comment (BevapDin) [inline src/cata_tiles.cpp]: This would look better as range based for loop:
```C++
for( auto &sprite : *g ) {
    if( sprite.weight == -1 ) {
    ...
}
```

Comment (sparr) [inline src/cata_tiles.cpp]: Thanks. I'm still new to C++11's new functionality in areas like this.

File: src/cata_tiles.h (hunk 2 above)
Comment (BevapDin) [inline src/cata_tiles.h]: > the default weight of 1 is functional for every case except the first one

Yes, so why are you initializing it to 1 anyway?

> on line 723 I'm creating an "invalid" entry 

I would assume that any default constructed `sprite_variation` is invalid (as it does not contain any sprites, the vector is empty). I see several places that test for `weight == -1`, a default constructed object would not fit there, but should still be considered invalid.

In short: this is inconsistent. If `weight == -1` means the object is invalid and `weight > 0` means it's valid, it should start that way.

**Comment by chaosvolt** [comment]
Hmm. This could lead to oddities like player constructions coming pre-graffitied, and I suspect it wouldn't work well for random monster/NPC appearances.

File: src/cata_tiles.h (hunk 2 above)
Comment (sparr) [inline src/cata_tiles.h]: "empty" and "invalid" are two different things. empty lists can get populated during the loading process. and items can be removed.

the -1 "invalid" marker is for exactly one case, the first variation in each list of variations. I *could* just start all of the loops at 1, skipping the 0th element, and then not bother with a weight on that variation, but that seems weirder.

File: src/cata_tiles.cpp (hunk 3 above)
Comment (BevapDin) [inline src/cata_tiles.cpp]: In my opinion, throwing an exception is better. If the user has written something wrong, they can easily correct it. And "users" are prone to ignored any kind of warning.

Comment (sparr) [inline src/cata_tiles.cpp]: the "user" here is the player, not the tileset author. I do not want players to get kicked out of the game because they tried to load an invalid tileset.

Comment (sparr) [inline src/cata_tiles.cpp]: or maybe I don't even want the tileset loading to fail. I want it to continue, and let the player play, and they will only see the warning if they don't like the results and go looking for it.

File: src/cata_tiles.h [hunk 8]
```diff
@@ -25,9 +25,37 @@ class JsonObject;
...
+        sprite_ids.push_back(id);
+    }
+};
+
 struct tile_type
 {
-    std::vector<int> fg, bg;
+    // first entry of each of these vectors is not a real sprite_variation
+    // but a bookkeeping list instead, used for reducing the task of selecting
+    // a random variation from O(N) to O(1)
```
Comment (BevapDin) [inline src/cata_tiles.h]: It would be better (if possible) to move this special part into another member of the struct. This kind of special entry is fragile and not easy to grasp.

File: src/cata_tiles.cpp [hunk 9]
```diff
@@ -1212,13 +1279,34 @@ bool cata_tiles::draw_sprite_at(const std::vector<int> &spritelist, int x, int y
...
+    int variation = 1;
+
+    if ( svlist.size() > 2 ) {
+        int lottery = loc_rand * (svlist[0].sprite_ids.size());
+        variation = svlist[0].sprite_ids[lottery];
+        // printf("Choosing among %d variations, with loc_rand %f lottery %d, chose %d\n",svlist.size()-1,loc_rand,lottery,variation);
+    }
+
+    // get sprite list from chosen variation
+    std::vector<int> spritelist = svlist[variation].sprite_ids;
```
Comment (BevapDin) [inline src/cata_tiles.cpp]: This is never written to, it should probably be a const reference, not a copy.

File: src/cata_tiles.cpp (hunk 3 above)
Comment (BevapDin) [inline src/cata_tiles.cpp]: > the "user" here is the player, not the tileset author. I do not want players to get kicked out of the game because they tried to load an invalid tileset.

I would have assumed that a tileset author tries their own tileset at least once. That's when the error message comes up and forces them to fix it. Anthe incorrect tileset won't even get published in the first place.

I can, however, understand your reasoning.

File: src/cata_tiles.h (hunk 8 above)
Comment (sparr) [inline src/cata_tiles.h]: the special part only appears in one such struct, out of a list of structs. if I moved it out, it would have to go adjacent to the list of structs, adding something like `vector<int> fg_pick, bg_pick` to tile_type. I considered that, but didn't like that concept.

using a class like weighted_list would resolve this problem, but as I mentioned elsewhere I'd have to improve that class first, and I didn't want to build such a class entirely from scratch here.

File: src/cata_tiles.cpp (hunk 3 above)
Comment (sparr) [inline src/cata_tiles.cpp]: There are currently open issues about json formatting errors that drive the game crazy, based on real json errors that made it through a PR into master. So, yeah, incorrect [everything] can get published :(

**Comment by sparr** [comment]
@chaosvolt player construction would definitely be odd, and might need to be excluded somehow as well. even if we assume the player is using salvaged materials (which isn't the case if they chop down trees themselves), it would be weird for the graffiti to follow the map square, not the material.

Re monsters/NPCs, I think that would be great. Attach the randomization to their ID, then we can have many different versions of the same zombie. There's no reason for every standard-zombie to be wearing the same color shirt, or have both arms, etc.

File: src/cata_tiles.cpp [hunk 10]
```diff
@@ -680,43 +716,72 @@ tile_type &cata_tiles::load_tile(JsonObject &entry, const std::string &id, int o
...
+        if ( entry.has_array(g.first) ) {
+            JsonArray g_array = entry.get_array(g.first);
+            // int elements of array indicates rotations
+            // create one variation, populate sprite_ids with list of ints
+            if ( g_array.test_int() ) {
+                g.second->push_back(sprite_variation());
+                while (g_array.has_more()) {
+                    const int s = g_array.next_int();
+                    if ( s >= 0 ) {
+                        g.second->at(0).sprite_ids.push_back(s);
```
Comment (BevapDin) [inline src/cata_tiles.cpp]: This is a result of the clever loop: Understanding what `g.second->at(0).sprite_ids` refers to is difficult. Apparently it's `curr_subtile.fg[0].sprite_ids` in one iteration and `curr_subtile.bg[0].sprite_ids` in the other?

At least extract `*g.second` to a named reference:
```C++
std::vector<sprite_variation> &variations = *g.second;
...
variations[0].sprite_ids.push_back(...)
```

Comment (sparr) [inline src/cata_tiles.cpp]: good argument in favor of function usage, or at least a named reference.

also, there's probably a bug here, related to which fg/bg element to write to. probably caused by me being confused by the deep nesting you're complaining about here.

File: src/cata_tiles.h (hunk 8 above)
Comment (BevapDin) [inline src/cata_tiles.h]: > adding something like  vector<int> fg_pick, bg_pick

Yes, logically separated data belongs into separate members.

It looks like `fg[0]` doesn't actually store tile-ids (as all the other elements of `fg` do), but indexes into the sprites vectors or so, is that about right? In that case, it should certainly be a separate member, as it is a completely different type of data. It's just a coincidence that you can use the same C++ type (vector) for it.

This allows you to get rid of the special case for `weight == -1` (for example in `load_tilejson_from_file`). It's much easier to read `curr_tile.fg_bookkeeping.push_back(s)` instead of `curr_tile.fg[0].push_back(s)`. ("Huh, what's so special about the first element here. Why are some of the loops in `load_tilejson_from_file` going over *all* elements, it doesn't seem special there.")

----

I don't get the last sentence. You *did* build the  sprite_variation class from scratch.

You should be able to extend the existing `weighted_list` class for your needs. You probably need to add the bookkeeping vector, filled either by overriding the `add` function or by a dedicated finalization function. The `get` function can be overridden similar to use the bookkeeping data.

Comment (sparr) [inline src/cata_tiles.h]: the annoying part of putting fg_bookkeeping outside of fg is that then every time I pass fg to a function (and that happens in a lot of places) I have to pass in fg_bookkeeping, too. that's why I used the approach that I did. and, again, a good reason to use the weighted_list class.

sprite_variation was "just" a struct at first. the initializers came later.

I'm thinking a finalizing function, so that the bookkeeping is optional.

Comment (BevapDin) [inline src/cata_tiles.h]: Put `fg` and bookkeeping thing into another class. You could even make this new class derive from `std::vector<int>`, so existing code that expects and reads a `std::vector<int>` will continue to work.

Actually this could be quite useful as `fg` and `bg` seem to have the same structure, the new class could be used for both.

Comment (sparr) [inline src/cata_tiles.h]: I'm just gonna improve weighted_list and use that. Solves or mootifies at least half of your objections above (all valid).

**Comment by sparr** [comment]
I think my latest commit hits most of your concerns, @BevapDin.

**Comment by Chezzo** [comment]
Hooray! Oh man, this is awesome.

**Comment by Chezzo** [comment]
>A tileset with real variant sprites would be a good thing to include with deployment of this patch.

Here ya go! This set changes the main male character and npc sprites to 5 different sprites, each set to 2 weight.

http://chezzo.com/cdda/ChestHole32TilesetOnRandom.zip

**Comment by sparr** [comment]
@Chezzo ha, thanks! now I really do need to add in proper handling for players/monsters in this patch.

File: src/weighted_list.h [hunk 11]
```diff
@@ -17,37 +17,43 @@ template <typename W, typename T> struct weighted_list {
     weighted_list() : total_weight(0) { };
 
     /**
-     * This will add a new object to the weighted list.
+     * This will add a new object to the weighted list. Returns a pointer to
+       the added or updated object, or NULL if weight was zero.
```
Comment (BevapDin) [inline src/weighted_list.h]: The documentation is a bit misleading. If weight is 0 (or less), it will return a pointer to the last element, or invoke undefined behavior if the list is currently empty. It will never (reliably) return NULL.

File: src/weighted_list.h [hunk 12]
```diff
@@ -124,41 +137,87 @@ template <typename W, typename T> struct weighted_list {
...
+    void precalc();
+
 protected:
     W total_weight;
-    std::vector<weighted_object<W, T> > objects;
+    weighted_object_list objects;
 
-    virtual size_t pick_ent() const =0;
+    virtual size_t pick_ent(long long) const =0;
+    void invalidate_precalc() {}
```
Comment (BevapDin) [inline src/weighted_list.h]: This should probably be virtual.

File: src/weighted_list.h [hunk 13]
```diff
@@ -124,41 +137,87 @@ template <typename W, typename T> struct weighted_list {
...
-        int picked = rng(1, this->total_weight);
-        int accumulated_weight = 0;
+    size_t pick_ent(long long rand_ll) const override {
         size_t i;
-        for(i=0; i < this->objects.size(); i++) {
-            accumulated_weight += this->objects[i].weight;
-            if(accumulated_weight >= picked) {
-                break;
+        int picked;
+        picked = (rand_ll%(this->total_weight))+1;
```
Comment (BevapDin) [inline src/weighted_list.h]: Combine this and the line above into one `int picked = ...;` - no point in having those as separate statements.

Btw. if you make `precalc_array` mutable, you can change it in this (const) function. Which means you can re-calculated it here, when you need it for the first time. Consequently, you don't need the `precalc` function. It would also allow you to bypass it when the list contains only one entry.

Just a sketch:
```C++
{
    if( objects.size() == 1 ) {
        return 0;
    }
    if( precalcs.empty() ) {
        // fill precalcs with proper values
    }
    return precalcs[...]
}
```

File: src/cata_tiles.cpp [hunk 14]
```diff
@@ -1228,13 +1327,25 @@ bool cata_tiles::draw_sprite_at(const std::vector<int> &spritelist, int x, int y
     destination.w = tile_width;
     destination.h = tile_height;
 
+    if ( svlist.empty() ) {
+        // render nothing
+        return true;
+    }
+
+    const std::vector<int> *spritelist = svlist.pick(loc_rand);
```
Comment (BevapDin) [inline src/cata_tiles.cpp]: You don't check for NULL here, which is fine as the loading makes the sure each tile has at least one sprite list, so `pick` should never return NULL. But should assign the result to a reference. This allows to get rid of the changed member access: `.foo` to `->foo`.

Additionally: what's with all the `at` calls? I see no catch of `std:.out_of_range`, and the index is checked anyway, so it should never throw anyway. Why not use the `operator[]`?

File: src/cata_tiles.cpp [hunk 15]
```diff
@@ -1213,13 +1268,57 @@ bool cata_tiles::draw_from_id_string(std::string id, TILE_CATEGORY category,
...
+        case C_WEATHER:
+            // TODO come up with ways to make random sprites consistent for these types
+            break;
+        case C_MONSTER:
+            // monsters (and player?), seed with index into monster list
+            // TODO detect player character, seed on name?
+            // FIXME add persistent id to Creature ttripoint(x,y,z_coord)ype, instead of using monster list index
+            seed = g->mon_at(tripoint(x,y,z_coord));
+            break;
+        default:
```
Comment (BevapDin) [inline src/cata_tiles.cpp]: Please don't add a default case when switching of an enum value. The compiler will warn you if the switch does not handle a value from the enumeration (for example when somebody added a new value, but forgot to update this switch). The default case disables this feature.

File: src/cata_tiles.cpp [hunk 16]
```diff
@@ -1213,13 +1268,57 @@ bool cata_tiles::draw_from_id_string(std::string id, TILE_CATEGORY category,
...
+        case C_TRAP:
+        case C_NONE:
+        case C_BULLET:
+        case C_HIT_ENTITY:
+        case C_WEATHER:
+            // TODO come up with ways to make random sprites consistent for these types
+            break;
+        case C_MONSTER:
+            // monsters (and player?), seed with index into monster list
+            // TODO detect player character, seed on name?
```
Comment (BevapDin) [inline src/cata_tiles.cpp]: Player character and NPCs have a unique id, see `player::getId()`. Monsters haven't.

File: src/cata_tiles.cpp [hunk 17]
```diff
@@ -594,40 +621,40 @@ void cata_tiles::load_ascii_set(JsonObject &entry, int offset, int size)
         id[7] = static_cast<char>(FG);
         id[8] = static_cast<char>(-1);
         tile_type &curr_tile = tile_ids[id];
-        curr_tile.fg.push_back(index_in_image + offset);
+        std::vector<int> *sprites = curr_tile.fg.add(std::vector<int>({index_in_image + offset}),1);
```
Comment (BevapDin) [inline src/cata_tiles.cpp]: Similar to below, if you're not checking for null anyway, make it a reference and use `operator[]`.

File: src/weighted_list.h (hunk 13 above)
Comment (sparr) [inline src/weighted_list.h]: for now, precalc is intentionally optional. making it unconditionally calculate on first pick_ent would be a bigger change, and deserve its own PR with comment from the folks responsible for the other code that uses weighted_list.

File: src/cata_tiles.cpp (hunk 14 above)
Comment (sparr) [inline src/cata_tiles.cpp]: I'm not a fan of using auto so much when it doesn't significantly improve readability (such as in range loops). As I've mentioned in our previous discussions, it makes it very hard for someone new to the codebase to figure out the type of things and follow the code.

The at() calls were because I couldn't figure out the right syntax to use operator[] on a pointer to a vector. Switching to a reference fixes that, thanks.

File: src/cata_tiles.cpp (hunk 15 above)
Comment (sparr) [inline src/cata_tiles.cpp]: That's a neat feature, thanks!

File: src/cata_tiles.cpp (hunk 16 above)
Comment (sparr) [inline src/cata_tiles.cpp]: There's a good chance I'm going to move that functionality up to Creature so Monsters will have it too. However, that's another patch for another day.

If I wanted to do it here, how would you suggest I differentiate between players/npcs and monsters here?

**Comment by sparr** [comment]
@Chezzo think you could throw together a tileset with a few different trees or grass? something a little more immediately visible in a default game.

File: src/cata_tiles.cpp (hunk 14 above)
Comment (BevapDin) [inline src/cata_tiles.cpp]: This is not the least about auto, it's about pointers vs references. My example only had auto as I was to lazy to type the types.

> figure out the right syntax to use operator[] on a pointer to a vector

`x = (*some_pointer)[index];` or `x = some_pointer->operator[](index);`

**Comment by Chezzo** [comment]
You got it @sparr! 

This set has the same random as above, with additional random grass and dirt tiles and some random trees. 

http://chezzo.com/cdda/ChestHole32TilesetRandomTreesNGrass.zip

**Comment by sparr** [comment]
> It's a cold spring in New England.

Don't even get me started on seasonal sprites...

**Comment by sparr** [comment]
@Chezzo my game complains about illegal trailing commas in your tile_config.json

I fixed that, filtered out a few other changes, changed the weights so there's less litter, and committed your tileset instead of the retrodays hack I was using.

**Comment by chaosvolt** [comment]
I'm going to hope that @sparr was actually satisfied with the PR and ready for merging? Because merging a PR that still has [WIP] in the title does not bode well.

**Comment by sparr** [comment]
kevin checked in with me on IRC, prompting me to make that last commit, and then him to merge.

**Comment by chaosvolt** [comment]
Ah, that explains it. Sorry. ^^"

**Comment by SeanMirrsen** [comment]
A hypothetical question, could this be used to create randomized terrain in non-tiles graphics modes? Like terrain in Dwarf Fortress, for instance. Grass and dirt look way too uniform currently.

**Comment by kevingranade** [comment]
The same logic definitely could, it would just need to be factored out a bit and applied to the curses drawing code.
